engines/scumm/gfx_gui.cpp

```diff
@@ -1964,7 +1991,163 @@ void ScummEngine::showMainMenu() {
+		break;
+	case GUI_CTRL_RESTART_BUTTON:
+		_menuPage = GUI_PAGE_RESTART;
+		setUpMainMenuControls();
+		drawMainMenuControls();
+		ScummEngine::drawDirtyScreenParts();
+		break;
+	case GUI_CTRL_OK_BUTTON:
+		if (_menuPage == GUI_PAGE_RESTART) {
+			_cursor.state = 0;
+			CursorMan.showMouse(false);
+			hasLoadedState = true;
+			restart();
+			return true;
+		} else if (_menuPage == GUI_PAGE_CODE_CONFIRM) {
+			_bootParam = atoi(_mainMenuSegaCDPasscode);
+
+			// We are running script 61 twice:
+			// - The first time we are providing it the passcode as an argument;
+			// - The second time we give no arguments, allowing var 63 to be updated.
+			// This will let us know whether we have successfully loaded a game or not.
+
+			// First time...
+			int args[16];
+			memset(args, 0, sizeof(args));
+			args[0] = _bootParam;
+
+			runScript(61, 0, 0, args);
+
+			// Second time...
+			args[0] = 0;
+			runScript(61, 0, 0, args);
+
+			hasLoadedState = _scummVars[63] == _bootParam;
+
+			_bootParam = 0;
+
+			if (!hasLoadedState) {
+				_menuPage = GUI_PAGE_INVALID_CODE;
+				setUpMainMenuControls();
+				drawMainMenuControls();
+				updateMainMenuControls();
+				ScummEngine::drawDirtyScreenParts();
+
+				waitForTimer(420);
+
+				_menuPage = GUI_PAGE_MAIN;
+				setUpMainMenuControls();
+				drawMainMenuControls();
+				updateMainMenuControls();
+				ScummEngine::drawDirtyScreenParts();
+			} else {
+				return true;
+			}
+		}
+
+		break;
+	case GUI_CTRL_CANCEL_BUTTON:
+		_menuPage = GUI_PAGE_MAIN;
+		setUpMainMenuControls();
+		drawMainMenuControls();
+		updateMainMenuControls();
+		ScummEngine::drawDirtyScreenParts();
+		break;
+	case GUI_CTRL_ARROW_LEFT_BUTTON:
+	case GUI_CTRL_ARROW_RIGHT_BUTTON:
+		if (_menuPage == GUI_PAGE_MAIN) {
+			if (op == GUI_CTRL_ARROW_LEFT_BUTTON) {
+				_defaultTextSpeed = CLIP<int>(_defaultTextSpeed - 1, 0, 9);
+			} else {
+				_defaultTextSpeed = CLIP<int>(_defaultTextSpeed + 1, 0, 9);
+			}
+
+			ConfMan.setInt("original_gui_text_speed", _defaultTextSpeed);
+			setTalkSpeed(_defaultTextSpeed);
+			syncSoundSettings();
+			ConfMan.flushToDisk();
+			drawMainMenuControls();
+			updateMainMenuControls();
+			ScummEngine::drawDirtyScreenParts();
+		}
+		break;
+	case GUI_CTRL_TEXT_SPEED_SLIDER:
+		_defaultTextSpeed = CLIP<int>((mouseX - 150) / 9, 0, 9);
+
+		ConfMan.setInt("original_gui_text_speed", _defaultTextSpeed);
+		setTalkSpeed(_defaultTextSpeed);
+		syncSoundSettings();
+		ConfMan.flushToDisk();
+		updateMainMenuControls();
+		ScummEngine::drawDirtyScreenParts();
+		break;
+	case GUI_CTRL_NUMPAD_0:
+	case GUI_CTRL_NUMPAD_1:
+	case GUI_CTRL_NUMPAD_2:
+	case GUI_CTRL_NUMPAD_3:
+	case GUI_CTRL_NUMPAD_4:
+	case GUI_CTRL_NUMPAD_5:
+	case GUI_CTRL_NUMPAD_6:
+	case GUI_CTRL_NUMPAD_7:
+	case GUI_CTRL_NUMPAD_8:
+	case GUI_CTRL_NUMPAD_9:
+	case GUI_CTRL_NUMPAD_BACK:
+	{
+		int inputNum = op == GUI_CTRL_NUMPAD_0 ? 0 : op;
+		uint curIdx;
+		for (curIdx = 0; curIdx < sizeof(_mainMenuSegaCDPasscode); curIdx++) {
```

Comment:
could it be 
```C
curIdx = strnlen(_mainMenuSegaCDPasscode, sizeof(_mainMenuSegaCDPasscode));
```

or even `strlen()` as you have 8 as max length anyway.

Reply:
Definitely! Thanks!